File: robot_navigation/scripts/distance_calculator.py
#!/usr/bin/env python

import rospy
from sensor_msgs.msg import LaserScan
from geometry_msgs.msg import Point
from math import sin,cos

class ObstacleDistanceCalculator():
    def __init__(self):
        rospy.init_node("obstacle_distance_calculator")
        self.obstacle_distance =Point()
        self.obstacle_radius = 0.3  # Define the obstacle radius in meters
        self.min_distance = float("inf")  # Initialize to positive infinity

        # Subscribe to LiDAR data
        rospy.Subscriber("/scan", LaserScan, self.lidar_callback)

        # Publish the obstacle distance
        self.obstacle_distance_pub = rospy.Publisher("/landmark_distance", Point, queue_size=10)

    def lidar_callback(self, data):
        # Find the minimum distance in the LiDAR scan within the obstacle radius
        for i, distance in enumerate(data.ranges):
            if data.range_min <= distance <= data.range_max and distance < self.min_distance:
                self.min_distance = distance

        # Calculate x and y distance from the robot
        angle = data.angle_min + (i * data.angle_increment)
        self.obstacle_distance.x = self.min_distance * cos(angle)
        self.obstacle_distance.y = self.min_distance * sin(angle)
        # Publish the obstacle distance
        self.obstacle_distance_pub.publish(self.obstacle_distance)

    def run(self):
        rate = rospy.Rate(10)  # 10 Hz
        while not rospy.is_shutdown():
            rate.sleep()

if __name__ == "__main__":
    try:
        obstacle_calculator = ObstacleDistanceCalculator()
        obstacle_calculator.run()
    except rospy.ROSInterruptException:
        pass

File: robot_navigation/scripts/distance_calculator.cpp
#include <ros/ros.h>
#include <sensor_msgs/LaserScan.h>
#include <geometry_msgs/Point.h>
#include <cmath>

class ObstacleDistanceCalculator {
public:
    ObstacleDistanceCalculator() : nh("~") {
        nh.param("obstacle_radius", obstacle_radius, 0.3);

        min_distance = std::numeric_limits<double>::infinity();

        lidar_sub = nh.subscribe("/scan", 1, &ObstacleDistanceCalculator::lidarCallback, this);
        obstacle_distance_pub = nh.advertise<geometry_msgs::Point>("/landmark_distance", 1);
    }

    void lidarCallback(const sensor_msgs::LaserScan::ConstPtr& scan) {
        min_distance = std::numeric_limits<double>::infinity();
        size_t min_index = 0;

        for (size_t i = 0; i < scan->ranges.size(); ++i) {
            if (scan->range_min <= scan->ranges[i] && scan->ranges[i] <= scan->range_max && scan->ranges[i] < min_distance) {
                min_distance = scan->ranges[i];
                min_index = i;
            }
        }

        if (std::isfinite(min_distance)) {
            double angle = scan->angle_min + min_index * scan->angle_increment;

            geometry_msgs::Point obstacle_distance;
            obstacle_distance.x = min_distance * cos(angle);
            obstacle_distance.y = min_distance * sin(angle);

            obstacle_distance_pub.publish(obstacle_distance);
        }
    }

    void run() {
        ros::Rate rate(10);  // 10 Hz
        while (ros::ok()) {
            ros::spinOnce();
            rate.sleep();
        }
    }

private:
    ros::NodeHandle nh;
    ros::Subscriber lidar_sub;
    ros::Publisher obstacle_distance_pub;
    double obstacle_radius;
    double min_distance;
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "obstacle_distance_calculator");

    ObstacleDistanceCalculator calculator;
    calculator.run();

    return 0;
}
